src/js/components/Upload.jsx

```diff
@@ -24,66 +56,178 @@ class Upload extends Component {
   static propTypes = {
     dialogOpen: PropTypes.bool.isRequired,
     dispatch: PropTypes.func.isRequired,
-    uploadDataTypes: PropTypes.object.isRequired
+    newSource: PropTypes.object.isRequired,
+    source: PropTypes.string.isRequired,
+    sources: PropTypes.object.isRequired,
+    uploadData: PropTypes.object.isRequired,
+    uploadDataTypes: PropTypes.object.isRequired,
+    uploads: PropTypes.object.isRequired
   }
 
   constructor (props) {
     super(props)
 
-    this.parseComplete = ::this.parseComplete
-    this.parseError = ::this.parseError
+    this.displayName = 'Upload'
+
+    this.newSourceParseComplete = ::this.newSourceParseComplete
+    this.newSourceParseError = ::this.newSourceParseError
+    this.onChangeExistingSource = ::this.onChangeExistingSource
+    this.onCreateNewSource = ::this.onCreateNewSource
+    this.onExistingSourceFileChange = ::this.onExistingSourceFileChange
+    this.onExistingSourceFileReject = ::this.onExistingSourceFileReject
+    this.onNewSourceFileChange = ::this.onNewSourceFileChange
+    this.onNewSourceFileReject = ::this.onNewSourceFileReject
+    this.onNewSourceNameBlur = ::this.onNewSourceNameBlur
+    this.onNewSourceNameChange = ::this.onNewSourceNameChange
     this.shouldRejectFile = ::this.shouldRejectFile
-    this.onFileChange = ::this.onFileChange
-    this.onFileReject = ::this.onFileReject
+    this.submitExistingSourceData = ::this.submitExistingSourceData
+    this.submitNewSourceData = ::this.submitNewSourceData
 
     this.state = {open: false}
   }
 
+  componentWillMount () {
+    const {dispatch} = this.props
+
+    dispatch(fetchSources())
+    dispatch(resetSource())
+  }
+
   componentWillUnmount () {
     const {dispatch} = this.props
 
     dispatch(resetUploadDataTypes())
+    dispatch(setExistingSourceData({}))
+    dispatch(setNewSourceData({}))
+    dispatch(setSource(''))
+    dispatch(setExistingSourceSelected(false))
   }
 
   shouldRejectFile (file) {
     return getFileExtension(file.name) !== 'csv'
   }
-
-  onFileReject (file) {
+  
+  existingSourceParseComplete (results, file) {
     const {dispatch} = this.props
+    const headers = Object.keys(results.data[0])
 
-    dispatch(resetUploadDataTypes())
-    dispatch(toggleDialog(true))
-  }
-
-  onFileChange (file) {
-    Papa.parse(file, {
-      complete: this.parseComplete,
-      dynamicTyping: true,
-      encoding: 'utf-8',
-      error: this.parseError,
-      preview: 2
-    })
+    dispatch(setExistingSourceData(JSON.stringify(results.data)))
+    
+    if (headers.length === 0) {
+      dispatch(disableNewSourceDataSubmit())
+    } else {
+      dispatch(enableNewSourceDataSubmit())
+    }
   }
 
-  parseComplete (results, file) {
+  newSourceParseComplete (results, file) {
     const {dispatch} = this.props
-    const headers = results.data[0]
-    const values = results.data[1]
+    const headers = Object.keys(results.data[0])
```

Comment:
I don't know if results.data[0] can ever be null, but I would probably use destructuring of results.data and just give it a default of an empty array. It might also be better to put the check for headers.length === 0 right here.